experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java

```
@@ -412,6 +415,46 @@ public ExpRunImpl createExperimentRun(Container container, String name)
     }
 
+    @Override
+    public ExpRun createRunForProvenanceRecording(Container container, User user, RecordedActionSet actionSet, String runName, @Nullable Integer runJobId)
+    {
+        try
+        {
+            ExpProtocol protocol;
+            try (DbScope.Transaction transaction = ExperimentService.get().getSchema().getScope().ensureTransaction(ExperimentService.get().getProtocolImportLock()))
+            {
+                List<String> sequenceProtocols = new ArrayList<>();
+                Map<String, ExpProtocol> protocolCache = new HashMap<>();
+                for (RecordedAction ra : actionSet.getActions())
+                {
+                    String stepName = ra.getName();
+                    sequenceProtocols.add(stepName);
+                    if (!protocolCache.containsKey(stepName))
+                    {
+                        // Check if it's in the database already
+                        ExpProtocol stepProtocol = ExperimentService.get().getExpProtocol(container, stepName);
+                        if (stepProtocol == null)
+                        {
+                            stepProtocol = ExperimentService.get().createExpProtocol(container, ExperimentRun, stepName);
+                            stepProtocol.save(user);
+                        }
+                        protocolCache.put(stepName, stepProtocol);
+                    }
+                }
+                Lsid provenanceLsid = new Lsid(PROVENANCE_PROTOCOL_LSID + container.getEntityId().toString());
+                protocol = ExpGeneratorHelper.ensureProtocol(container, user, protocolCache, sequenceProtocols, provenanceLsid, runName, LOG);
+
+                transaction.commit();
+            }
+            return ExpGeneratorHelper.insertRun(container, user, actionSet, runName, runJobId, protocol, null, null, null);
+        }
+        catch (ExperimentException | ValidationException e)
+        {
+            LOG.error(e);
+        }
+        return null;
```

Generally, better to rethrow the exception instead of returning null. Is there an expected error case where we want to fail and return null?

---

done